generator/wrapper_gen.py

```diff
@@ -175,9 +175,13 @@ def my_procid():
 
 def _get_data(x):
     if isinstance(x, pdDataFrame):
-        x = [x.loc[:,i].values for i in x]
+        x_dtypes = x.dtypes.values
+        if np.all(x_dtypes == x_dtypes[0]):
+            x = x.to_numpy()
+        else:
+            x = [xi.to_numpy() for _, xi in x.items()]
```

Comment:
I would still use `np.ascontiguousarray(xi.to_numpy())` just for safety sake. 

It won't make a copy if the array is already contiguous, but will safe-guard the code against possible future changes in pandas.

Reply:
For some reason, there is an overhead for calling this function. But it seems that DataFrames do not allow such a case. I added three tests and the results are correct. Here is one example:

```python
        X = np.random.randn(13024*3, 16)
        df = pd.DataFrame(X[1::3, 1:])
        k = 2
        dir_dtypes = {col: 'float64' if i % 2 == 0 else 'float32' for i, col in enumerate(df.columns)}
        df = df.astype(dir_dtypes)
        ps = pd.Series(X[1::3, 0])
        df['newcol'] = ps
        self.verify_on_dbscan(df)
```

Reply:
I strongly recommend introducing a (cheap) check which raises an error if we don't meet what we expect. Changes that introduce a copy should not be able to pass unnoticed.

Reply:
Add check in function `_get_data`?

Reply:
Instead of `np.ascontiguousarray(xi.to_numpy())` I suggest adding the check and then using just `xi.to_numpy()`.

Reply:
The `np.ascontiguous` is about 10 times slower than checking flags:

```
In [37]: x = np.random.randn(10**5, 50)

In [38]: %timeit y = np.ascontiguousarray(x)
814 ns ± 7.84 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [39]: %timeit y2 = x
17.8 ns ± 0.062 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [43]: %timeit x.flags['C']
76.9 ns ± 2.06 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Raising an error is not very user friendly, but if we decide on it, it is most performant to do in `make_nt` where we can do it in C.

Reply:
Emitting a mutable warning (that a copy was made) would also be acceptable to me.

Reply:
OK. I added the following check to the code. Look please.

```c++

    else if (PyList_Check(obj) && PyList_Size(obj) > 0)
        { // a list of arrays for SOA?
               ....

                for (auto i = 0; i < N; i++)
                {
                    PyArrayObject * ary = (PyArrayObject *)PyList_GetItem(obj, i);
                   ....

                    if (!(array_is_behaved(ary) || array_is_behaved_F(ary)))
                    {
                        std::cerr << "Can't get the data correctly without a copy. Errors are possible.\n";
                    }

```

p.s. Sorry for the big changes in cpp. They were made automatically by the clang-format.

Reply:
The `*ary`  should be a 1D array, and so `array_is_behaves(ary)` should be sufficient.

When an automatic tool makes changes like that, it is best if automatic changes are separated from meaningful changes into a separate commit.

This makes it easy to opt-out of these changes if needed.

Reply:
@oleksandr-pavlyk Yes. You're right. I fixed.